File: sangsoo/0x0E Sort1/mergesort.cpp
#include <vector>
#include <iostream>

using namespace std;

void merge(vector<int>& L, int first, int mid, int last){
    int k = first;
    vector<int> sub1(L.begin()+first, L.begin()+mid+1);
    vector<int> sub2(L.begin()+mid+1, L.begin()+last+1);

    int i = 0, j = 0;
    while(i < sub1.size() && j < sub2.size()){
        if(sub1[i] <= sub2[j]){
            L[k] = sub1[i];
            i++;
        }
        else{
            L[k] = sub2[j];
            j++;
        }
        k++;
    }
    while (i < sub1.size()){
        L[k++] = sub1[i++];
    }
    while (j < sub2.size()){
        L[k++] = sub2[j++];
    }
}

void mergeSortHelp(vector<int>& L, int first, int last){
    if(first == last){
        return;
    }
    else{
        int mid = first + (last-first) / 2;
        mergeSortHelp(L, first, mid);
        mergeSortHelp(L, mid+1, last);
        merge(L, first, mid, last);
    }
}

void mergeSort(vector<int>& L){
    mergeSortHelp(L, 0, L.size()-1);
}

int main(){
    vector<int> L = {-4, -102, 1, 3, 0, 2};
    mergeSort(L);
    for (auto num : L){
        cout << num << " ";
    }
    return 0;
}

File: sangsoo/0x0E Sort1/mergesort.py
# CFDS0, prof kim.

def merge(L:list, first:int, mid: int, last: int)->None:
    k = first
    sub1 = L[first:mid+1]
    sub2 = L[mid+1:last+1]
    i = j = 0
    while i  < len(sub1) and j < len(sub2):
        if sub1[i] <= sub2[j]:
            L[k] = sub1[i]
            i = i + 1
        else:
            L[k] = sub2[j]
            j = j + 1
        k = k + 1
    if i < len(sub1):
        L[k:last+1] = sub1[i:]
    elif j < len(sub2):
        L[k:last+1] = sub2[j:]


def mergeSortHelp(L:list, first:int, last:int)->None:
    if first == last: # Conditional statement
        return # Base case
    else:
        mid = first + (last - first)//2
        mergeSortHelp(L, first, mid) # recursive call for sublist1
        mergeSortHelp(L, mid+1, last) # recursive call for sublist2
        merge(L, first, mid, last)

def mergeSort(L:list)->None:
    mergeSortHelp(L, 0, len(L)-1)

if __name__ == "__main__":

    L = [5, 2, 3, 4, 1]
    mergeSort(L)
    print(L)

